src/coreclr/jit/emit.h

```diff
@@ -2175,19 +2175,34 @@ class emitter
         return (emitCurIG && emitCurIGfreeNext > emitCurIGfreeBase);
     }
 
-    instrDesc* emitLastIns;
+#define EMIT_MAX_IG_INS_COUNT 256
+#define EMIT_MAX_LAST_INS_COUNT 32
+
+    // A recording of the last instructions, used for peephole optimizations.
+    instrDesc* emitLastInstrs[EMIT_MAX_LAST_INS_COUNT];
+    unsigned   emitCurLastInsCnt;
     insGroup*  emitLastInsIG;
 
+    inline instrDesc* emitGetLastIns() const
+    {
+        return emitLastInstrs[(emitCurLastInsCnt - 1) % ArrLen(emitLastInstrs)];
+    }
+
+    inline bool emitHasLastIns() const
+    {
+        return (emitCurLastInsCnt > 0);
+    }
+
     // Check if a peephole optimization involving emitLastIns is safe.
     //
     // We must have a non-null emitLastIns to consult.
     // The emitForceNewIG check here prevents peepholes from crossing nogc boundaries.
     // The final check prevents looking across an IG boundary unless we're in an extension IG.
-    bool emitCanPeepholeLastIns()
+    bool emitCanPeepholeLastIns() const
```

Comment:
What does it mean to check this function for the "last" instruction, when there are _N_ last instructions to check?

Also, there is a check on `emitLastInsIG`. But the `N` last instructions might be spread over multiple IGs.

Reply:
You think the check is not sufficient for checking back "N" instructions?

My line of thinking was, if I can peephole the last instruction, we should be able to check back further.

If we definitely cannot, then when we save the IG we need to reset the count if we know that it's not safe between IG boundaries, even if the IG is extended.

Reply:
> You think the check is not sufficient for checking back "N" instructions?

Well, if the "previous instruction list" crosses multiple IG boundaries, then I think this check would need to be written to take that into account -- maybe by checking for each instruction we look back.